djangocms_versioning/plugin_rendering.py

```diff
@@ -24,14 +27,30 @@ def prefetch_versioned_related_objects(instance, toolbar):
             qs = versionable.content_model.objects.all()
         related_field = getattr(instance, field.name)
         if related_field:
-            filters = {versionable.grouper_field_name: related_field}
+            is_related_manager = isinstance(related_field, models.Manager)
+
+            if is_related_manager:
+                filter_key = '{}__in'.format(versionable.grouper_field_name)
+                filter_values = related_field.all()
+                filters = {filter_key: filter_values}
+            else:
+                filters = {versionable.grouper_field_name: related_field}
             # TODO Figure out grouping values-awareness
             # for extra fields other than hardcoded 'language'
             if "language" in versionable.extra_grouping_fields:
                 filters["language"] = toolbar.request_language
             qs = qs.filter(**filters)
             prefetch_cache = {versionable.grouper_field.remote_field.name: qs}
-            related_field._prefetched_objects_cache = prefetch_cache
+
+            # TODO refine it after understand prefetch in many2many field.
+            # because if `related_field` is ManyRelatedManager, it is temporary,
+            # we can't store `_prefetched_objectes_cache` to it, so I decide to store the
+            # prefetched value to model instance.
```

Comment:
I'm thinking to prove this we will need to test how many queries are being performed without this and then with it to see if it actually does anything. I'm tempted to change candidate fields to ignore Many to Many relations.

We will need to create a performance test to prove this change to be sure that we are actually adding performance here and not  making it slower by performing a query that isn't actually used. 

self.assertNumQueries allows you to count how many queries are being performed. Cretae a page with many plugins, with and without M2M poll plugin and see if this value is actually changed by this change.

Reply:
@Aiky30 , @chaosk , @jonathan-s ,
I ever browser django cms source code but I haven't seen any place which possibly adopt this cache to improve the query performance. so Actually I don't understand why @chaosk 
 added this field-level prefetched_objects_cache.

Reply:
-@joshyu This is a Django feature called prefetch_related, that function uses prefetched_objects_cache for query speed ups. You should be searching Django not Django CMS: https://github.com/django/django/blob/93ed71e05802a47774b52503cdc3442686d686c1/django/db/models/base.py#L603

Reply:
@Aiky30 , @jonathan-s ,
I need some more time to finish this test.
I found some resources: 
https://medium.com/@lucasmagnum/djangotip-select-prefetch-related-e76b683aa457
https://github.com/django/django/blob/a34cb5a6d408203f4fbdb364fc9768c026eda224/django/db/models/fields/related_descriptors.py#L907

Reply:
@Aiky30  @jonathan-s ,
I update the test and now It can test the cache with 1000 relation access, but all hit the cache.